api/lib/models/project/project-data-mapper-impl.ts

```diff
@@ -0,0 +1,236 @@
+import { Inject, Injectable } from 'injection-js';
+import { from } from 'rxjs';
+import { map } from 'rxjs/operators';
+import * as squel from 'squel';
+import { UidType } from '../../JSCommon/common-types';
+import { HttpError } from '../../JSCommon/errors';
+import { AuthLevel } from '../../services/auth/auth-types';
+import { IAcl, IAclPerm } from '../../services/auth/RBAC/rbac-types';
+import { IDbResult } from '../../services/database';
+import { GenericDataMapper } from '../../services/database/svc/generic-data-mapper';
+import { MysqlUow } from '../../services/database/svc/mysql-uow.service';
+import { IUowOpts } from '../../services/database/svc/uow.service';
+import { Logger } from '../../services/logging/logging';
+import { IActiveHackathonDataMapper } from '../hackathon/active-hackathon';
+import { IProjectDataMapper } from './index';
+import { Project } from './project';
+
+@Injectable()
+export class ProjectDataMapperImpl extends GenericDataMapper
+  implements IAclPerm, IProjectDataMapper {
+
+  public COUNT: string = 'project:count';
+  public CREATE: string = 'project:create';
+  public DELETE: string = 'project:delete';
+  public READ: string = 'project:read';
+  public READ_ALL: string;
+  public UPDATE: string = 'project:update';
+
+  public tableName: string = 'PROJECT_LIST';
+  public teamTableName: string = 'PROJECT_TEAM';
+  protected pkColumnName: string = 'projectID';
+
+  constructor(
+    @Inject('IAcl') acl: IAcl,
+    @Inject('MysqlUow') protected readonly sql: MysqlUow,
+    @Inject('IActiveHackathonDataMapper') protected readonly activeHackathonDataMapper: IActiveHackathonDataMapper,
+    @Inject('BunyanLogger') protected readonly logger: Logger,
+  ) {
+    super(acl);
+    super.addRBAC(
+      [this.DELETE],
+      [AuthLevel.TECHNOLOGY],
+      undefined,
+      [AuthLevel[AuthLevel.DIRECTOR]],
+    );
+    super.addRBAC(
+      [this.COUNT],
+      [AuthLevel.TEAM_MEMBER],
+      undefined,
+      [AuthLevel[AuthLevel.VOLUNTEER]],
+    );
+    super.addRBAC(
+      [this.READ_ALL],
+      [AuthLevel.VOLUNTEER],
+      undefined,
+      [AuthLevel[AuthLevel.PARTICIPANT]],
+    );
+    super.addRBAC(
+      [this.CREATE, this.READ, this.UPDATE],
+      [AuthLevel.PARTICIPANT],
+    );
+  }
+
+  public delete(projectID: UidType): Promise<IDbResult<void>> {
```

Comment:
Based on a change on the main branch, pls add this TODO.
```suggestion
  // TODO: Change parameter to CompoundUidType
  public delete(projectID: UidType): Promise<IDbResult<void>> {
```